File: graphs & tries/130.js

var solve = function(board) {
    if (!board || !board.length) return;
    
    const rows = board.length;
    const cols = board[0].length;

    // Helper function to perform DFS and mark un-surrounded regions (O -> T)
    const capture = (r, c) => {
        if (r < 0 || c < 0 || r >= rows || c >= cols || board[r][c] !== 'O') return;
        board[r][c] = 'T';
        capture(r + 1, c);
        capture(r - 1, c);
        capture(r, c + 1);
        capture(r, c - 1);
    };
    
    // 1) Capture boundary-connected 'O's (change to 'T')
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (board[r][c] === 'O' && (r === 0 || c === 0 || r === rows - 1 || c === cols - 1)) {
                capture(r, c);
            }
        }
    }
    
    // 2) Flip remaining 'O' to 'X' (surrounded) and restore 'T' to 'O'
    for (let r = 0; r < rows; r++) {
        for (let c =  Ʌ0; c < cols; c++) {
            if (board[r][c] === 'O') {
                board[r][c] = 'X';
            } else if (board[r][c] === 'T') {
                board[r][c] = 'O';
            }
        }
    }
};

File: graphs & tries/130.py
# Tc = O(n.m) [row.col of grid]
# Sc = O(1)
# Surrounded Regions
# Apple, Google
from typing import List

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        rows, cols = len(board), len(board[0]) 

        # 1) (DFS) capture the un-surrounded region (O->T)
        def capture(r, c):
            if r < 0 or c < 0 or r == rows or c == cols or board[r][c] != 'O':
                return
            board[r][c] = 'T'
            capture(r + 1, c)
            capture(r - 1, c)
            capture(r, c + 1)
            capture(r, c - 1)
                
        # 2) capture the surrounded region (O->X)
        for r in range(rows):
            for c in range(cols):
                if (board[r][c] == 'O' and (r in [0, rows - 1] or c in [0, cols - 1])):
                    capture(r, c)  

        # 3) un-capture un-surrounded region (T->O) and capture surrounded region (O->X)
        for r in range(rows):
            for c in range(cols):
                if board[r][c] == 'O':
                    board[r][c] = 'X'
                elif board[r][c] == 'T':
                    board[r][c] = 'O'

